tests/integration/test_nginx_route.py

```diff
@@ -0,0 +1,192 @@
+    """Requests get, but will retry when the response status code is not 200."""
+    time_start = time.time()
+    while True:
+        response = requests.get(url, headers={"Host": host_header}, timeout=5)
+        if response.status_code == 200 or time.time() - time_start > retry_timeout:
+            return response
+
+
+@pytest_asyncio.fixture(scope="module")
+async def build_and_deploy(ops_test: OpsTest, run_action):
+    """build and deploy nginx-ingress-integrator charm.
+
+    Also deploy and relate an any-charm application for test purposes.
+
+    Returns: None.
+    """
+
+    assert isinstance(ops_test.model, Model)
+
+    async def build_and_deploy_ingress():
+        charm = await ops_test.build_charm(".")
+        assert isinstance(ops_test.model, Model)
+        return await ops_test.model.deploy(
+            str(charm), application_name="ingress", series="focal", trust=True
+        )
+
+    await asyncio.gather(
+        ops_test.model.deploy(
+            "any-charm",
+            application_name="any",
+            channel="beta",
+            config={"src-overwrite": gen_src_overwrite()},
+        ),
+        build_and_deploy_ingress(),
+    )
+    await ops_test.model.wait_for_idle()
+    await run_action(ANY_APP_NAME, "rpc", method="start_server")
+    relation_name = f"{INGRESS_APP_NAME}:nginx-route"
+    await ops_test.model.add_relation(f"{ANY_APP_NAME}:nginx-route", relation_name)
+    await ops_test.model.wait_for_idle()
+
+
+@pytest.mark.usefixtures("build_and_deploy")
+async def test_ingress_connectivity():
+    """
+    arrange: given charm has been built and deployed.
+    act: access ingress IP address with correct host name in HTTP headers.
+    assert: HTTP request should be forwarded to the application.
+    """
+    response = requests_get("http://127.0.0.1/ok", host_header="any")
+
+    assert response.text == "ok"
+    assert response.status_code == 200
+    assert (
+        requests_get("http://127.0.0.1/ok", host_header="any.other", retry_timeout=0).status_code
+        == 404
+    )
+
+
+@pytest.mark.usefixtures("build_and_deploy")
+async def test_update_service_hostname(ops_test: OpsTest):
+    """
+    arrange: given charm has been built and deployed.
+    act: update the service-hostname option in any-charm.
+    assert: HTTP request with the service-hostname value as the host header should be forwarded
+        to the application correctly.
+    """
+    new_hostname = "any.other"
+    assert ops_test.model
+    await ops_test.model.applications[ANY_APP_NAME].set_config(
+        {"src-overwrite": gen_src_overwrite(service_hostname=new_hostname)}
+    )
+    await ops_test.model.wait_for_idle(status="active")
+
+    response = requests_get("http://127.0.0.1/ok", host_header=new_hostname)
+    assert response.text == "ok"
+    assert response.status_code == 200
+
+    await ops_test.model.applications[ANY_APP_NAME].set_config(
```

Comment:
Is this cleanup? Should it be done in a fixture?

Reply:
I can't put this in a fixture since I have to do some tests after the reset of configuration.

Reply:
Which tests? It looks like this is the last part of the test? If the fixture is function scoped the cleanup should get executed before the next tests

Reply:
I thinks in the `test_update_additional_hosts` there are some test following the removal of the configuration field to test if corresponding field is removed in the relation data.

Reply:
That indicates that the tests could be relying on execution order here? In that case it would be especially good to move the setup and teardown to a fixture so that a test could be run in isolation

Reply:
I don't think so, the configuration is reset in the same test case.